Can't use constants in enum definition

### Describe the bug

Hello,

It seems that constants can not be used in enum definition... yet ?

Kind regards,

### Expected Behavior

Output:
```
example
```

### Current Behavior

Output:

```
main.v:7:12: error: the default value for an enum has to be an integer
    5 | 
    6 | enum Test {
    7 |     example = first_const
      |               ~~~~~~~~~~~
    8 | }
    9 |
```

When cast to int:
```
==================
/tmp/v_1000/main.18018808916242146123.tmp.c:815: error: '_const_main__first_const' undeclared
...
==================
```

### Reproduction Steps

Run the following file:

```
const (
	first_const = 1000
	second_const = 2000
)

enum Test {
	example = first_const
}

fn main() {
	println(Test.example)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.0 b06811c.4355be8

### Environment details (OS name and version, etc.)

V full version: V 0.4.0 b06811c.4355be8
OS: linux, Debian GNU/Linux 12 (bookworm)
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz

getwd: /home/lenain/Documents/Projects/V/test/src
vexe: /home/lenain/.bin/v/v
vexe mtime: 2023-08-11 14:10:10

vroot: OK, value: /home/lenain/.bin/v
VMODULES: OK, value: /home/lenain/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.39.2
Git vroot status: weekly.2023.32-23-g4355be8e
.git/config present: true

CC version: cc (Debian 12.2.0-14) 12.2.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1818,8 +1818,11 @@ fn (mut c Checker) enum_decl(mut node ast.EnumDecl) {
 						if field.expr.language == .c {
 							continue
 						}
-						if field.expr.kind in [.constant, .unresolved] {
+						if field.expr.kind == .unresolved {
 							c.ident(mut field.expr)
+						}
+						if field.expr.kind == .constant && field.expr.obj.typ.is_int() {
+							/* accepts int constants as enum value */
 							continue
 						}
 					}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1818,6 +1818,10 @@ fn (mut c Checker) enum_decl(mut node ast.EnumDecl) {
 						if field.expr.language == .c {
 							continue
 						}
+						if field.expr.kind in [.constant, .unresolved] {
+							c.ident(mut field.expr)
+							continue
+						}
 					}
 					mut pos := field.expr.pos()
 					if pos.pos == 0 {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3875,7 +3875,16 @@ fn (mut g Gen) enum_decl(node ast.EnumDecl) {
 		if field.has_expr {
 			g.enum_typedefs.write_string(' = ')
 			expr_str := g.expr_string(field.expr)
-			g.enum_typedefs.write_string(expr_str)
+			if field.expr is ast.Ident && field.expr.kind == .constant {
+				const_def := g.global_const_defs[util.no_dots(field.expr.name)]
+				if const_def.def.starts_with('#define') {
+					g.enum_typedefs.write_string(const_def.def.all_after_last(' '))
+				} else {
+					g.enum_typedefs.write_string(expr_str)
+				}
+			} else {
+				g.enum_typedefs.write_string(expr_str)
+			}
 			cur_enum_expr = expr_str
 			cur_enum_offset = 0
 		} else if is_flag {

--- a/vlib/v/tests/const_as_enum_value_test.v
+++ b/vlib/v/tests/const_as_enum_value_test.v
@@ -1,0 +1,15 @@
+const (
+	first_const  = 1000
+	second_const = 2000
+)
+
+enum Test {
+	example = first_const
+}
+
+fn test_main() {
+	a := Test.example
+	dump(Test.example)
+	assert a == unsafe { Test(1000) }
+	assert Test.example == unsafe { Test(1000) }
+}

